### ISSUE TITLE
Optional one-to-one and sonata_type_admin

### ISSUE THREAD

**Issue by debach** [sonata-project/SonataAdminBundle#3554]
I have an _optional_ one-to-one relationship (say, a _Person_ might have one or none _Certificate_) and want to embed a form for the inverse side (Certificate) into the form for the owning side (Person) with `sonata_type_admin`. Certificate has a couple of required and a couple of optional fields.

If I go ahead and use

``` php
->add('certificate', 'sonata_type_admin')
```

then all requried fields of Certificate will be required in the form and I will be unable to submit the form without filling the required fields for Certificate. If I add `'required' => false` and leave all fields empty, Sonata will still attempt to create a Certificate. A workaround is to overwrite the `prePersist()` and `preUpdate()` methods to unset the inverse side if some of its required fields are empty, but that’s bad for a couple of reasons.

It would be nice if `'required' => false` added a checkbox to the form – like “Add Certificate?” – and Sonata only created and linked a Certificate to the Person if the checkbox was checked.

See also [Stop SonataAdmin / Symfony2 from creating empty objects with sonata_type_admin embedded admins](https://stackoverflow.com/q/18157485/1231686).

**Comment by metamaker** [comment]
Same here. Creating empty objects with sonata_type_admin embedded admins is really disaster if you have onPersist/onRemove callbacks.

**Comment by aurelijusrozenas** [comment]
+1

**Comment by ph538834** [comment]
+1

**Comment by core23** [comment]
Instead of posting stupid +1 comments, could someone create a PR? 

A simple +1 would create an auto PR by some magic fairy

**Comment by jordisala1991** [comment]
Probably using a ChoiceFieldMaskType could be a good solution for this, no need to implement anything more. I just need to check what happens when the field is hidden.

Leaving this opened for now, but will close once I confirm that it works.

**Comment by debach** [comment]
I stopped using Sonata a while ago and haven’t tried the proposal with ChoiceFieldMaskType. The documentation doesn’t state whether fields hidden by ChoiceFieldMaskType are still marked as required or not. Also, deletion of the linked entity likely isn’t handled by ChoiceFieldMaskType either.

What I did eventually is create a custom type for optional one-to-one relationships, which is embedded in the owning side form. If there is a non-null linked inverse entity, its edit form is shown along with a "delete" checkbox. Otherwise, a "create" checkbox is shown. This approach didn’t require a lot of code, but was more complicated than I anticipated (custom `data_class`, Mapper and Transformer).

Let me know if you’re interested in any details.
